File: distributed-job-manager/src/test/java/ru/fix/distributed/job/manager/strategy/AssignmentStrategyUtils.java
package ru.fix.distributed.job.manager.strategy;

import ru.fix.distributed.job.manager.model.JobId;
import ru.fix.distributed.job.manager.model.WorkItem;
import ru.fix.distributed.job.manager.model.WorkerId;
import ru.fix.distributed.job.manager.model.AssignmentState;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class AssignmentStrategyUtils {

    static void addWorkerWithItems(AssignmentState state, String worker, int workItemsCount, int jobsCount) {
        HashSet<WorkItem> workItems = new HashSet<>();

        for (int i = 0; i < workItemsCount; i++) {
            for (int j = 0; j < jobsCount; j++) {
                workItems.add(new WorkItem("work-item-" + i,  new JobId("job-" + j)));
            }
        }
        state.addWorkItems(new WorkerId(worker), workItems);
    }

    static Map<JobId, Set<WorkerId>> generateAvailability(AssignmentState assignmentState) {
        Map<JobId, Set<WorkerId>> availability = new HashMap<>();

        for (Map.Entry<WorkerId, HashSet<WorkItem>> workerEntry : assignmentState.entrySet()) {
            for (WorkItem workItem : workerEntry.getValue()) {
                availability.computeIfAbsent(workItem.getJobId(), state -> new HashSet<>())
                        .add(workerEntry.getKey());
            }
        }

        return availability;
    }

    static HashSet<WorkItem> generateItemsToAssign(AssignmentState assignmentState) {
        HashSet<WorkItem> itemsToAssign = new HashSet<>();

        for (Map.Entry<WorkerId, HashSet<WorkItem>> workerEntry : assignmentState.entrySet()) {
            itemsToAssign.addAll(workerEntry.getValue());
        }

        return itemsToAssign;
    }

    static int calculateReassignments(AssignmentState stateBefore, AssignmentState stateAfter) {
        int count = 0;

        for (Map.Entry<WorkerId, HashSet<WorkItem>> stateEntry : stateBefore.entrySet()) {
            WorkerId workerId = stateEntry.getKey();

            for (WorkItem workItem : stateEntry.getValue()) {
                if (!stateAfter.containsWorkItemOnWorker(workerId, workItem)) {
                    count++;
                }
            }
        }
        return count;
    }

    static Set<WorkItem> generateWorkItems(JobId jobId, int indexFromInclusive, int indexToExclusive) {
        return IntStream.range(indexFromInclusive, indexToExclusive)
                .mapToObj(index -> new WorkItem("work-item-" + index, jobId) )
                .collect(Collectors.toSet());
    }
}

File: distributed-job-manager/src/test/kotlin/ru/fix/distributed/job/manager/strategy/AssignmentStrategyUtils.kt
package ru.fix.distributed.job.manager.strategy

import ru.fix.distributed.job.manager.JobId
import ru.fix.distributed.job.manager.model.*

class WorkerScope(private val state: AssignmentState) {
    operator fun String.invoke(builder: JobScope.() -> Unit) {
        builder(JobScope(state, this))
    }
}

class JobScope(
        private val state: AssignmentState,
        private val worker: String
) {
    operator fun String.invoke(vararg items: String) {
        for (item in items) {
            state.addWorkItem(WorkerId(worker), WorkItem(item, JobId(this)))
        }
    }
}

fun assignmentState(builder: WorkerScope.() -> Unit) =
        AssignmentState().apply {
            builder(WorkerScope(this))
        }

class AvailabilityScope(private val availability: Availability){
    operator fun String.invoke(vararg items: String) {
        availability[JobId(this)] = items.map { WorkerId(it) }.toHashSet()
    }
}
fun availability(builder: AvailabilityScope.() -> Unit) =
        Availability().apply {
            builder(AvailabilityScope(this))
        }

class WorkItemScope(private val workItems: HashSet<WorkItem>){
    operator fun String.invoke(vararg items: String) {
        for(item in items){
            workItems.add(WorkItem(item, JobId(this)))
        }
    }
}
fun workItems(builder: WorkItemScope.() -> Unit) =
        HashSet<WorkItem>().apply {
            WorkItemScope(this).builder()
        }

fun generateAvailability(assignmentState: AssignmentState): Availability {
    val availability = Availability()

    for ((key, value) in assignmentState) {
        for (workItem in value) {
            availability.getOrPut(workItem.jobId) { hashSetOf<WorkerId>() }.add(key)
        }
    }
    return availability
}

fun generateItemsToAssign(assignmentState: AssignmentState): HashSet<WorkItem> {
    val itemsToAssign = HashSet<WorkItem>()

    for ((_, value) in assignmentState) {
        itemsToAssign.addAll(value)
    }
    return itemsToAssign
}

fun calculateReassignments(stateBefore: AssignmentState, stateAfter: AssignmentState): Int {
    var reassignments = 0

    for ((workerId, workItems) in stateAfter) {
        for (workItem in workItems) {
            if (!stateBefore.containsWorkItemOnWorker(workerId, workItem)) {
                reassignments++
            }
        }
    }
    return reassignments
}

fun AssignmentState.getLocalWorkPoolSizeInfo(availability: Map<JobId, Set<WorkerId>>): String {
    val info = StringBuilder("Work pool size per job:\n")
    availability.forEach { (jobId: JobId?, availableWorkers: Set<WorkerId?>) ->
        info.append(jobId)
                .append(" - work pool size: ")
                .append(localPoolSize(jobId))
                .append("\n")
        availableWorkers.forEach { workerId: WorkerId? ->
            info.append("\t")
                    .append(workerId)
                    .append(": ")
                    .append(this.getWorkItems(workerId, jobId).size)
                    .append("\n")
        }

    }
    return info.toString()
}

fun AssignmentState.getGlobalWorkPoolSizeInfo(): String? {
    val info = StringBuilder("Global load per worker:\n")
    this.forEach { workerId: WorkerId?, items: HashSet<WorkItem?> ->
        info.append("\t")
                .append(workerId)
                .append(": ")
                .append(items.size)
                .append("\n")
    }
    return info.toString()
}

class Report(
        private val availability: Map<JobId, Set<WorkerId>>? = null,
        private val itemsToAssign: Set<WorkItem>? = null,
        private val prevAssignment: AssignmentState? = null,
        private val newAssignment: AssignmentState? = null
) {

    override fun toString(): String {
        return "".plus(availability?.let { availability(availability) } ?: "")
                .plus(itemsToAssign?.let { itemsToAssign(itemsToAssign) } ?: "")
                .plus(prevAssignment?.let { "Previous $prevAssignment" } ?: "")
                .plus(newAssignment?.let { "New $newAssignment" } ?: "")
    }

    private fun itemsToAssign(itemsToAssign: Set<WorkItem>): String {
        val jobs = HashMap<JobId, MutableSet<WorkItem>>()
        itemsToAssign.forEach { item ->
            jobs.getOrPut(item.jobId) { mutableSetOf() }.add(item)
        }

        val picture = StringBuilder("Items to assign:\n")
        jobs.forEach { (jobId, workItems) ->
            picture.append("\t└ ${jobId.id}\n ")
            workItems.forEach { workItem -> picture.append("\t\t└ ${workItem.id}\n") }
        }
        return picture.toString()
    }

    private fun availability(availability: Map<JobId, Set<WorkerId>>): String {
        val picture = StringBuilder("Availability:\n")

        availability.forEach { (jobId, workerIds) ->
            picture.append("\t└ ${jobId.id}\n")
            workerIds.forEach { workerId -> picture.append("\t\t└ ${workerId.id}\n") }
        }
        return picture.toString()
    }
}

